utils/compare-output/check.py

```diff
@@ -0,0 +1,178 @@
+
+            try:
+                if parts and not parts[0].startswith('#'):
+                    if len(parts) != 2:
+                        raise ValueError('invalid number of fields')
+
+                    try:
+                        key = int(parts[0])
+                    except:
+                        raise ValueError('failed to parse vertex id '
+                                '("{}" is not a number)'
+                                .format(parts[0]))
+
+                    try:
+                        value = value_parser(parts[1])
+                    except:
+                        raise ValueError('failed to parse vertex value '
+                                '("{}" is not a value)'
+                                .format(parts[1]))
+
+                    if key in results:
+                        raise ValueError('duplicate entry for vertex {}'
+                                .format(key))
+
+                    results[key] = value
+
+            except ValueError as e:
+                if verbose:
+                    print('error in {} on line {}: {}'.format(filename,
+                        index + 1, e), file=sys.stderr)
+
+                errors += 1
+
+    return results, errors
+
+
+def check_exact():
+    return lambda x, y: x == y
+
+
+def check_threshold():
+    def comparison(x, y):
+        global COMPARISON_THRESHOLD
+
+        # Check exact match first to handle +/-inf
+        return x == y or abs(x-y) < COMPARISON_THRESHOLD
+
+    return comparison
+
+
+def check_equivalence():
+    equiv_x2y = {}
+    equiv_y2x = {}
+
+    def comparison(x, y):
+        if x not in equiv_x2y:
+            equiv_x2y[x] = y
+
+        if y not in equiv_y2x:
+            equiv_y2x[y] = x
+
+        return equiv_a2b[x] == y and equiv_b2a[y] == x
+
+    return comparison
+
+
+def compare_results(file_a, results_a, file_b, results_b, comparison, verbose):
+    keys = set()
+    keys.update(results_a.keys())
+    keys.update(results_b.keys())
+
+    errors = 0
+
+    for key in keys:
+        try:
+            if key not in results_a:
+                raise ValueError('vertex id {} is missing in {}'.format(key, file_a))
+
+            if key not in results_b:
+                raise ValueError('vertex id {} not found in {}'.format(key, file_b))
+
+            if not comparison(results_a[key], results_b[key]):
+                raise ValueError('{} does not match {} for vertex id {}'.format(
+                    results_a[key], results_b[key], key))
+
+        except ValueError as e:
+            if verbose:
+                print('error comparing {} to {}: {}'.format(file_a, file_b,
+                    e), file=sys.stderr)
+            errors += 1
+
+    return len(keys) - errors, errors
+
+
+def get_algorithm_params(algorithm):
+    if algorithm in ('bfs'):
+        return int, check_exact
+    elif algorithm in ('pr', 'lcc', 'sssp'):
+        return float, check_threshold
+    elif algorithm in ('wcc', 'cdlp'):
```

Comment:
Can we not compare CDLP with `check_exact`? Every vertex should start with its own ID as its label, so I do not see a reason to check for equivalence instead of an exact match.

For WCC it does not really matter because the component ID does not influence the outcome of the algorithm.